frontend/src/store/modules/sheet.ts

```diff
@@ -0,0 +1,207 @@
+  setSheetList: typeof mutations.SET_SHEET_LIST;
+};
+
+const actions = {
+  ...makeActions<ActionsMap>({
+    setSheetState: types.SET_SHEET_STATE,
+    setSheetList: types.SET_SHEET_LIST,
+  }),
+  // create
+  async createSheet(
+    { commit, state, rootState, rootGetters }: any,
+    sheetRecord?: CreateSheetState
+  ): Promise<Sheet> {
+    const ctx = rootState.sqlEditor.connectionContext as ConnectionContext;
+    const instance = rootGetters["instance/instanceById"](ctx.instanceId);
+    const database = rootGetters["database/databaseById"](ctx.databaseId);
+    const tab = rootGetters["tab/currentTab"] as TabInfo;
+
+    const result = (
+      await axios.post(`/api/sheet`, {
+        data: {
+          type: "createSheet",
+          attributes: {
+            instanceId: ctx.instanceId,
+            databaseId: ctx.databaseId,
+            name: tab.label,
+            statement: tab.queryStatement,
+            visibility: "PRIVATE",
+          },
+        },
+      })
+    ).data;
+    const newSheet = convertSheet(result.data, result.included);
+    newSheet.instance = instance;
+    newSheet.database = database;
+
+    commit(
+      types.SET_SHEET_LIST,
+      (state.sheetList as Sheet[])
+        .concat(newSheet)
+        .sort((a, b) => b.createdTs - a.createdTs)
+    );
+
+    commit(types.SET_SHEET_BY_ID, newSheet);
+
+    return newSheet;
+  },
+  // retrieve
+  async fetchSheetList({ commit, dispatch, state }: any) {
+    dispatch("setSheetState", { isFetchingSheet: true });
+
+    const data = (await axios.get(`/api/sheet`)).data;
+    const sheetList: Sheet[] = data.data.map((sheet: ResourceObject) => {
+      const newSheet = convertSheet(sheet, data.included);
+      commit(types.SET_SHEET_BY_ID, newSheet);
+      return newSheet;
+    });
+
+    commit(
+      types.SET_SHEET_LIST,
+      sheetList.sort((a, b) => b.createdTs - a.createdTs)
+    );
+
+    dispatch("setSheetState", { isFetchingSheet: false });
+  },
+  // update
+  async patchSheetById(
+    { dispatch }: any,
+    { id, name, statement, visibility }: Partial<Sheet>
+  ): Promise<Sheet> {
+    const attributes: Partial<
+      Pick<Sheet, "name" | "statement" | "visibility">
+    > = {};
+    if (name) {
+      attributes.name = name;
+    }
+    if (statement) {
+      attributes.statement = statement;
+    }
+    if (visibility) {
+      attributes.visibility = visibility;
+    }
+
+    const result = (
+      await axios.patch(`/api/sheet/${id}`, {
+        data: {
+          type: "sheetPatch",
+          attributes,
+        },
+      })
+    ).data;
+
+    const newSheet = convertSheet(result.data, result.included);
+
+    dispatch("fetchSheetList");
+
+    return newSheet;
+  },
+  // delete
+  async deleteSheet({ commit, state }: any, id: number) {
+    const sheet = state.sheetById.get(id);
+
+    await axios.delete(`/api/sheet/${id}`);
+    commit(types.REMOVE_SHEET, sheet);
+  },
+  // upsert
+  async upsertSheet(
```

Comment:
So that i can use a single function to update the sheet store. with or without id.

Reply:
IMO, that's less readable and error prone because:

1. We lose the type checking because of using Partial
2. It makes hard to understand the code, normally, the caller should have exact knowledge of whether to use create or patch.

All in all,  upsert has different semantics from create and patch. And we'd better to use it if we DO need its unique semantic, instead of just using it as a syntax sugar for convenience.